server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java

```diff
@@ -165,6 +172,30 @@ public void testInternalSendExceptionForksToGenericIfHandlerDoesNotFork() {
 
             assertEquals("simulated exception in sendRequest", getSendRequestException(future, IOException.class).getMessage());
         }
+        assertWarnings(
+            "[transport.enable_stack_protection] setting was deprecated in Elasticsearch and will be removed in a future release."
+        );
+    }
+
+    public void testInternalSendExceptionWithNonForkingResponseHandlerCanCompleteFutureWaiterFromGenericThread() throws Exception {
+        try (var nodeA = new TestNode("node-A")) {
+            final var future = new PlainActionFuture<TransportResponse.Empty>();
+            nodeA.transportService.getThreadPool().generic().execute(() -> {
+                assertEquals("simulated exception in sendRequest", getSendRequestException(future, IOException.class).getMessage());
+            });
+            final Thread callingThread = Thread.currentThread();
+            nodeA.transportService.sendRequest(
+                nodeA.getThrowingConnection(),
+                TestNode.randomActionName(random()),
+                new EmptyRequest(),
+                TransportRequestOptions.EMPTY,
+                new ActionListenerResponseHandler<>(future.delegateResponse((l, e) -> {
+                    assertThat(Thread.currentThread(), is(callingThread));
```

Comment:
nit: doesn't _really_ matter since `Thread.equals()` does the right thing, but I'd prefer `sameInstance`:

```suggestion
                    assertThat(Thread.currentThread(), sameInstance(callingThread));
```